_episodes/05-dimensionality-reduction.md

```diff
@@ -3,15 +3,156 @@ title: "Dimensionality Reduction"
 - "Understand that reducing the number of dimensions can simplify modelling and allow classifications to be performed."
 - "Recall that PCA is a popular technique for dimensionality reduction."
-- "Recall that TSNE is another technique for dimensionality reduction."
-- "Apply PCA and TSNE with Scikit Learn to an example dataset."
-- "Evaluate the relative peformance of PCA and TSNE."
+- "Recall that t-SNE is another technique for dimensionality reduction."
+- "Apply PCA and t-SNE with Scikit Learn to an example dataset."
+- "Evaluate the relative peformance of PCA and t-SNE."
 keypoints:
-- "PCA is a dimensionality reduction technique"
-- "TSNE is another dimensionality reduction technique"
+- "PCA is a linear dimensionality reduction technique for tabular data"
+- "t-SNE is another dimensionality reduction technique for tabular data that is more general than PCA"
 ---
+# Dimensionality Reduction
+
+Dimensionality reduction is the process of using a subset of the coordinates, 
+which may be transformed, of the dataset to capture the variation in features 
+of the data set.  It can be a helpful pre-processing step before doing other 
+operations on the data, such as classification, regression or visualization.
+
+## Dimensionality Reduction with Scikit-learn
+
+First setup our environment and load the MNIST digits dataset which will be used 
+as our initial example.
+
+~~~
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn import decomposition
+from sklearn import datasets
+from sklearn import manifold
+
+digits = datasets.load_digits()
+
+# Examine the dataset
+
+print(digits.data)
+
+print(digits.target)
+
+X = digits.data
+y = digits.target
+
+~~~
+{: .language-python}
+
+### Principle Component Analysis (PCA)
+
+PCA is a technique that does rotations of data in a two dimensional
+array to decompose the array into combinations vectors that are orthogonal
+and can be ordered according to the amount of information they carry.
+
+~~~
+# PCA
+pca = decomposition.PCA(n_components=2)
+pca.fit(X)
+X_pca = pca.transform(X)
+
+fig = plt.figure(1, figsize=(4, 4))
+plt.clf()
+plt.scatter(X_pca[:, 0], X_pca[:, 1], c=y, cmap=plt.cm.nipy_spectral, 
+        edgecolor='k',label=y)
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("pca.svg")
+~~~
+{: .language-python}
+
+![Reduction using PCA](../fig/pca.svg)
+
+### t-distributed Stochastic Neighbor Embedding (t-SNE)
+
+~~~
+# t-SNE embedding
+tsne = manifold.TSNE(n_components=2, init='pca',
+        random_state = 0)
+X_tsne = tsne.fit_transform(X)
+fig = plt.figure(1, figsize=(4, 4))
+plt.clf()
+plt.scatter(X_tsne[:, 0], X_tsne[:, 1], c=y, cmap=plt.cm.nipy_spectral,
+        edgecolor='k',label=y)
+plt.colorbar(boundaries=np.arange(11)-0.5).set_ticks(np.arange(10))
+plt.savefig("tsne.svg")
+~~~
+{: .language-python}
+
+![Reduction using t-SNE](../fig/tsne.svg)
+
+
+
+> # Exercise: Working in three dimensions
```

Comment:
I took a closer look and this line must be replaced with `> ## Exercise: Parameters`, that is, the header must have two `##` and not one `#`.

Reply:
ok